File: php/tests/LoggerTest.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq\Test;

require_once __DIR__ . '/TestRunner.php';
require_once __DIR__ . '/../Logger.php';

use Apache\Rocketmq\Logger;

class LoggerTest
{
    private $testLogFile;

    public function setUp(): void
    {
        $this->testLogFile = sys_get_temp_dir() . '/rocketmq_test_logger_' . uniqid() . '.log';
    }

    public function tearDown(): void
    {
        Logger::close();
        if (file_exists($this->testLogFile)) {
            @unlink($this->testLogFile);
        }
    }

    public function testSingleton()
    {
        Logger::setLogFile($this->testLogFile);
        $logger1 = Logger::getInstance('TestComponent');
        $logger2 = Logger::getInstance('TestComponent');

        TestRunner::assertTrue($logger1 === $logger2, "Should return same instance for same component");
    }

    public function testDifferentComponents()
    {
        Logger::setLogFile($this->testLogFile);

        $logger1 = Logger::getInstance('ComponentA');
        $logger2 = Logger::getInstance('ComponentB');

        TestRunner::assertTrue($logger1 !== $logger2, "Should return different instances for different components");
    }

    public function testLogLevelFiltering()
    {
        Logger::setLogFile($this->testLogFile);
        Logger::setLogLevel(Logger::LEVEL_ERROR);

        $logger = Logger::getInstance('LogLevelTest');
        $logger->debug('This should not be logged');
        $logger->info('This should not be logged');
        $logger->warning('This should not be logged');

        $content = '';
        if (file_exists($this->testLogFile)) {
            $content = file_get_contents($this->testLogFile) ?: '';
        }
        TestRunner::assertFalse(
            strpos($content, 'This should not be logged') !== false,
            "DEBUG/INFO/WARNING should be filtered when log level is ERROR"
        );
    }

    public function testLogLevelWrite()
    {
        Logger::setLogFile($this->testLogFile);
        Logger::setLogLevel(Logger::LEVEL_DEBUG);

        $logger = Logger::getInstance('LogLevelWriteTest');
        $marker = 'unique_test_marker_' . uniqid();
        $logger->info($marker);

        $content = file_get_contents($this->testLogFile) ?: '';
        TestRunner::assertTrue(
            strpos($content, $marker) !== false,
            "INFO message should be written when log level is DEBUG"
        );
    }

    public function testLogFormat()
    {
        Logger::setLogFile($this->testLogFile);
        Logger::setLogLevel(Logger::LEVEL_DEBUG);

        $marker = 'format_test_' . uniqid();
        $logger = Logger::getInstance('FormatTest');
        $logger->info($marker);

        $content = file_get_contents($this->testLogFile) ?: '';
        $expectedPattern = "/\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3}\] \[INFO\] \[FormatTest\] {$marker}/";
        TestRunner::assertTrue(
            preg_match($expectedPattern, $content) === 1,
            "Log line should match expected format"
        );
    }

    public function testTimezoneDetection()
    {
        $originalTz = date_default_timezone_get();

        // Force UTC to simulate PHP CLI default
        date_default_timezone_set('UTC');

        // Logger init should detect and correct system timezone
        Logger::getInstance('TimezoneTest');

        $currentTz = date_default_timezone_get();

        // Restore original timezone
        if ($originalTz !== 'UTC') {
            date_default_timezone_set($originalTz);
        }

        // If system timezone was detected (not UTC anymore), the test passes
        // If it's still UTC, that's also valid (system may genuinely be UTC)
        TestRunner::assertTrue(
            in_array($currentTz, timezone_identifiers_list(), true) || $currentTz === 'UTC',
            "Timezone should be a valid identifier after Logger init (got: {$currentTz})"
        );
    }
}

echo "=== LoggerTest ===\n";
TestRunner::run(new LoggerTest());

File: golang/logger_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golang

import (
	"sync"
	"testing"
)

type recordedLogEntry struct {
	level     string
	message   string
	keyValues []any
}

type recordingLoggerState struct {
	mu      sync.Mutex
	entries []recordedLogEntry
}

func (s *recordingLoggerState) append(entry recordedLogEntry) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries = append(s.entries, entry)
}

func (s *recordingLoggerState) find(message string) (recordedLogEntry, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, entry := range s.entries {
		if entry.message == message {
			return entry, true
		}
	}
	return recordedLogEntry{}, false
}

type recordingLogger struct {
	state     *recordingLoggerState
	keyValues []any
}

var _ Logger = (*recordingLogger)(nil)

func (l *recordingLogger) record(level string, message string, keyValues ...any) {
	allKeyValues := append([]any{}, l.keyValues...)
	allKeyValues = append(allKeyValues, keyValues...)
	l.state.append(recordedLogEntry{
		level:     level,
		message:   message,
		keyValues: allKeyValues,
	})
}

func (l *recordingLogger) Debug(message string, keyValues ...any) {
	l.record("debug", message, keyValues...)
}

func (l *recordingLogger) Info(message string, keyValues ...any) {
	l.record("info", message, keyValues...)
}

func (l *recordingLogger) Warn(message string, keyValues ...any) {
	l.record("warn", message, keyValues...)
}

func (l *recordingLogger) Error(message string, keyValues ...any) {
	l.record("error", message, keyValues...)
}

func (l *recordingLogger) With(keyValues ...any) Logger {
	allKeyValues := append([]any{}, l.keyValues...)
	allKeyValues = append(allKeyValues, keyValues...)
	return &recordingLogger{state: l.state, keyValues: allKeyValues}
}

func TestSetLoggerIsUsedByNewClients(t *testing.T) {
	state := &recordingLoggerState{}
	SetLogger(&recordingLogger{state: state})
	t.Cleanup(ResetLogger)

	config := &Config{
		Endpoint:      "127.0.0.1:8081",
		ConsumerGroup: "test-group",
	}
	producer, err := NewProducer(config)
	if err != nil {
		t.Fatalf("failed to create producer: %v", err)
	}
	consumer, err := NewSimpleConsumer(config)
	if err != nil {
		t.Fatalf("failed to create simple consumer: %v", err)
	}

	sugarBaseLogger.Infof("package log: %d", 1)
	defaultProducer := producer.(*defaultProducer)
	defaultConsumer := consumer.(*defaultSimpleConsumer)
	defaultProducer.cli.log.Info("producer log")
	defaultConsumer.cli.log.Warnf("consumer log: %s", "warning")

	assertRecordedLog(t, state, "package log: 1", "info", "")
	assertRecordedLog(t, state, "producer log", "info", defaultProducer.cli.GetClientID())
	assertRecordedLog(t, state, "consumer log: warning", "warn", defaultConsumer.cli.GetClientID())
}

func assertRecordedLog(t *testing.T, state *recordingLoggerState, message string, level string, clientID string) {
	t.Helper()
	entry, ok := state.find(message)
	if !ok {
		t.Fatalf("log entry %q was not recorded", message)
	}
	if entry.level != level {
		t.Fatalf("log entry %q has level %q, want %q", message, entry.level, level)
	}
	if clientID == "" {
		if len(entry.keyValues) != 0 {
			t.Fatalf("package log has unexpected key-values: %v", entry.keyValues)
		}
		return
	}
	if len(entry.keyValues) != 2 || entry.keyValues[0] != "client_id" || entry.keyValues[1] != clientID {
		t.Fatalf("log entry %q has key-values %v, want [client_id %s]", message, entry.keyValues, clientID)
	}
}
